### PR TITLE
Fix for shared axes diverging after setting tick markers

### PR DESCRIPTION
**Pull request by shaynair**
## PR Summary

This fixes issue #8946. 
It adds code to methods including `set_xticks()` and `set_yticks()`, which originally did not immediately update the shared axes. This led to axes having diverging `get_xlim()`, producing a wrong plot.

Example plot:

Associated code snippet to reproduce the above plot:
```python
import matplotlib.pyplot as plt
fig, (ax1, ax2) = plt.subplots(2, sharex=True)
ax1.plot(range(0, 10), range(0, 10))
ax2.plot(range(0, 10), range(0, 10))
plt.xticks(range(0, 30))
plt.show()
```

The two plots should be identical as they share the x-axis, but clearly aren't. This is a fix for this.

## PR Checklist

- [x] Has Pytest style unit tests
- [x] Code is PEP 8 compliant
- [x] New features are documented, with examples if plot related
- [x] Documentation is sphinx and numpydoc compliant
- [x] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [x] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/axes/_base.py [hunk 1]
```diff
@@ -3203,6 +3203,12 @@ def set_xticks(self, ticks, minor=False):
         """
         ret = self.xaxis.set_ticks(ticks, minor=minor)
         self.stale = True
+
+        g = self.get_shared_x_axes()
+        for ax in g.get_siblings(self):
+            ax.xaxis.set_ticks(ticks, minor=minor)
```
Comment (efiring) [inline lib/matplotlib/axes/_base.py]: Given how sharing works, I don't think you need all these "set_ticks" lines.  I suspect that simply setting 'ax.stale = True' is enough.  Would you try that, please?

Comment (shaynair) [reply lib/matplotlib/axes/_base.py]: I just tried it (commenting line 3209). Unfortunately doing that fails the automated test I created, so it wouldn't fix the issue.

Comment (efiring) [reply lib/matplotlib/axes/_base.py]: Thank you.  Still, I'm pretty sure that calling set_ticks is not the right thing to do here.  It is breaking part of the sharing connection by giving the axes different instances of the Locators. Instead, it looks like what is needed is synchronization of the view_interval after it was changed in the call to set_ticks.  Logically, it seems like sharing should include the viewLim directly, but it doesn't, so when the viewLim is changed, the other axes need to be notified.

**Comment by efiring** [comment]
I think you are on the right track, but the needed changes can be made in a much more compact way:

- In `axis.XAxis.set_view_interval`, add the 'emit' kwarg and the shared-axis updating logic.
- In `axes._base.Axes.set_xlim`, call the modified method above, thereby eliminating the duplication.
- The same for `YAxis`, of course.
- I think this implies that the "stale" setting also migrates into the `set_view_interval` method.

I'm wondering whether the `figure.draw_idle()` calls should be left as they are, or whether they should be changed to `figure.stale = True`.  I don't understand the "stale" machinery well enough.  @tacaswell?

**Comment by shaynair** [comment]
Thanks for the feedback @efiring, just a few concerns:

- Would the `auto` parameter also be delegated to `set_view_interval` since it also needs to be propagated to the shared axes?
- I'm having trouble making `Axis3D` compatible with the `set_view_interval` change. `Axis3D` uses `v_interval` whereas `Axes3D.set_xlim` uses `xy_viewLim.intervalx` which are different. How would I add the `emit` parameter to `Axis3D.set_view_interval` without it breaking previous test cases?

**Comment by shaynair** [comment]
Closed this due to the commit history being broken.

**Comment by shaynair** [comment]
This pull request is updated here: #10691

**Pull request closed without merging**